external/fv3fit/fv3fit/keras/_models/shared/sequences.py

```diff
@@ -37,37 +46,52 @@ def __getitem__(self, idx) -> Tuple[np.ndarray, np.ndarray]:
         return X, y
 
 
-class _XyMultiArraySequence(tf.keras.utils.Sequence):
+class XyMultiArraySequence(tf.keras.utils.Sequence):
     """
-    Wrapper object converting a sequence of batch datasets
-    to a sequence of tuples of input/output numpy arrays.
+    Wrapper object converting a sequence of unstacked batch datasets
+    to a stacked, shuffled sequence of tuples of input/output numpy arrays.
 
     These tuples contain one unpacked numpy array for each input/output,
-    in contrast to _XyArraySequence which is specialized to the case
-    of a single input/output of packed arrays.
+    in contrast to XyArraySequence which is specialized to the case
+    of a single input/output of packed arrays. This class also performs
+    the responsibilities of StackedBatches, unlike XyArraySequence
     """
 
     def __init__(
         self,
         X_names: Sequence[str],
         y_names: Sequence[str],
         dataset_sequence: Sequence[xr.Dataset],
+        unstacked_dims=fv3gfs.util.Z_DIMS,
+        n_halo: int = 0,
     ):
         self.X_names = X_names
         self.y_names = y_names
         self.dataset_sequence = dataset_sequence
+        self.n_halo = n_halo
+        self.unstacked_dims = unstacked_dims
 
     def __len__(self) -> int:
         return len(self.dataset_sequence)
 
     def __getitem__(self, idx) -> Tuple[np.ndarray, np.ndarray]:
         ds = self.dataset_sequence[idx]
-        X = tuple(ds[name].values for name in self.X_names)
-        y = tuple(ds[name].values for name in self.y_names)
+        X_y_datasets = [ds[self.X_names], ds[self.y_names]]
+        X_y_datasets[0] = append_halos(X_y_datasets[0], n_halo=self.n_halo)
+        for i in range(2):
+            X_y_datasets[i] = stack(
+                X_y_datasets[i], unstacked_dims=self.unstacked_dims
+            ).dropna(dim=SAMPLE_DIM_NAME)
+            X_y_datasets[i] = check_empty(X_y_datasets[i])
+            X_y_datasets[i] = preserve_samples_per_batch(X_y_datasets[i])
+        X_y_datasets = shuffled(np.random, X_y_datasets)
```

Comment:
I also can't call such a helper separately on X and y because they need to be shuffled identically, without doing something risky like passing an identical random seed. If a bug ever makes these non-identical, it would be very hard to track down.